### DISCUSSION TITLE
Is variance of List is documented correctly?

### DISCUSSION THREAD

**Discussion by Murtagy** [python/typing#1300] [Q&A] [answered]
https://mypy.readthedocs.io/en/stable/generics.html#variance-of-generic-types

Documentation states that lists are invariant, although they act as covariant.

I have reproduced the example of Managers and Employees the docs mentions
```
class Employee: pass
class Manager(Employee): pass
class Boss(Manager): pass

def salaries(
    staff: list[Manager],
    accountant: Callable[[Manager], int]
) -> list[int]:
    return [accountant(s) for s in staff]

def accountant_M(m: Manager) -> int:
    return 200
```
And here are the results:
```
salaries([Boss()], accountant_M)  # valid (most specific)
salaries([Manager()], accountant_M)  # valid (mid specific)
salaries([Employee()], accountant_M)  # invalid (least specific)
```
If the List was **invariant** I would expect only middle one to be valid, 
Both top and middle are valid, so I guess it acts as **covariant**

**Comment by TeamSpen210** [comment, marked as answer]
List is invariant, the reason this works is that type checkers are smart enough to do bidirectional inference to find a solution. What it's doing is casting `Boss` to `Manager` (effectively discarding the fact it's a subclass), making the list a `list[Manager]`. Adding some variable annotations can help demonstrate the behaviour:
```py
bosses: list[Boss] = [Boss()]
salaries(bosses, accountant_M)  # Fails as expected
secret_boss: Manager = Boss()  # Valid cast, it's a subclass.
salaries([secret_boss], accountant_M)  # This works, because it's just an Manager here.
```

**Reply by Murtagy** [reply]
Thanks, was trying to settle the terms in my head.

Interesting, wonder wether this is a bug or a feature.
Inline auto casts seem to make sense though for practical purposes.
